File: LeetCode/Tag/Breadth-first Search/cpp/133.clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
    vector<Node*> new_nodes;
    
public:
    
    Solution() {
        new_nodes.resize(101);
    }

    Node* cloneGraph(Node* node) {
        if (!node) return NULL;
        queue<Node*> q;
        q.push(node);

        new_nodes[node->val] = new Node(node->val);
        while (!q.empty()) {
            Node* curr = q.front();
            q.pop();
            for (Node* u: curr->neighbors) {
                if (!new_nodes[u->val]) {
                    q.push(u);
                    new_nodes[u->val] = new Node(u->val);
                }
                new_nodes[curr->val]->neighbors.push_back(new_nodes[u->val]);
            }
        }
        return new_nodes[node->val];
    }
};

File: LeetCode/Tag/Recursion/js/133.clone-graph.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */

/**
 * @param {Node} node
 * @return {Node}
 */
var cloneGraph = function(node) {
  if(!node) return null;

  const map = {};
  const clone = root => {
    if(map[root.val]){
      return map[root.val];
    }
    else{
      map[root.val] = new Node(root.val);
      map[root.val].neighbors = root.neighbors.map(n => clone(n));
      return map[root.val];
    }
  }

  return clone(node);
};
